File: util/templates/devices/Pgp2bAxi.py
d = { 'name': "Pgp2bAxi",
     'registers': [{ 'name'   : "CountReset",
                    'address' : "0x00",
                    'type'    : "Configuration" },
                   {'name'    : "ResetRx",
                    'address' : "0x01",
                    'type'    : "Configuration" },
                   {'name'    : "Flush",
                    'address' : "0x02",
                    'type'    : "Configuration" },
                   {'name'    : "Loopback",
                    'address' : "0x03",
                    'type'    : "Configuration" },
                   {'name'    : "LocData",
                    'address' : "0x04",
                    'type'    : "Configuration" },
                   {'name'    : "AutoStatus",
                    'address' : "0x05",
                    'type'    : "Configuration" },
                   {'name'    : "Status",
                    'address' : "0x08",
                    'type'    : "Status" },
                   {'name'    : "RxPhyReady",
                    'address' : "0x08",
                    'size'    : "1",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "TxPhyReady",
                    'address' : "0x08",
                    'size'    : "1",
                    'lsBit'   : "1",
                    'type'    : "Status" },
                   {'name'    : "RxLocalLinkReady",
                    'address' : "0x08",
                    'size'    : "2",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "RxRemLinkReady",
                    'address' : "0x08",
                    'size'    : "3",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "TxLinkReady",
                    'address' : "0x08",
                    'size'    : "4",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "RxLinkPolarity",
                    'address' : "0x08",
                    'size'    : "2",
                    'lsBit'   : "8",
                    'type'    : "Status" },
                   {'name'    : "RxRemPause",
                    'address' : "0x08",
                    'size'    : "4",
                    'lsBit'   : "12",
                    'type'    : "Status" },
                   {'name'    : "TxLocPause",
                    'address' : "0x08",
                    'size'    : "4",
                    'lsBit'   : "16",
                    'type'    : "Status" },
                   {'name'    : "RxRemOverflow",
                    'address' : "0x08",
                    'size'    : "4",
                    'lsBit'   : "20",
                    'type'    : "Status" },
                   {'name'    : "TxLocOverflow",
                    'address' : "0x08",
                    'size'    : "4",
                    'lsBit'   : "24",
                    'type'    : "Status" },
                   {'name'    : "RxRemLinkData",
                    'address' : "0x09",
                    'type'    : "Status" },
                   {'name'    : "RxCellErrorCount",
                    'address' : "0x0A",
                    'type'    : "Status" },
                   {'name'    : "RxLinkDownCount",
                    'address' : "0x0B",
                    'type'    : "Status" },
                   {'name'    : "RxLinkErrorCount",
                    'address' : "0x0C",
                    'type'    : "Status" },
                   {'name'    : "RxRemOverflow0Count",
                    'address' : "0x0D",
                    'type'    : "Status" },
                   {'name'    : "RxRemOverflow1Count",
                    'address' : "0x0E",
                    'type'    : "Status" },
                   {'name'    : "RxRemOverflow2Count",
                    'address' : "0x0F",
                    'type'    : "Status" },
                   {'name'    : "RxRemOverflow3Count",
                    'address' : "0x10",
                    'type'    : "Status" },
                   {'name'    : "RxFrameErrorCount",
                    'address' : "0x11",
                    'type'    : "Status" },
                   {'name'    : "RxFrameCount",
                    'address' : "0x12",
                    'type'    : "Status" },
                   {'name'    : "TxLocOverflow0Count",
                    'address' : "0x13",
                    'type'    : "Status" },
                   {'name'    : "TxLocOverflow1Count",
                    'address' : "0x14",
                    'type'    : "Status" },
                   {'name'    : "TxLocOverflow2Count",
                    'address' : "0x15",
                    'type'    : "Status" },
                   {'name'    : "TxLocOverflow3Count",
                    'address' : "0x16",
                    'type'    : "Status" },
                   {'name'    : "TxFrameErrorCount",
                    'address' : "0x17",
                    'type'    : "Status" },
                   {'name'    : "TxFrameCount",
                    'address' : "0x18",
                    'type'    : "Status" },
                   {'name'    : "RxClkFreq",
                    'address' : "0x19",
                    'type'    : "Status" },
                   {'name'    : "TxClkFreq",
                    'address' : "0x1A",
                    'type'    : "Status" },
                   {'name'    : "TxOpCodeStatus",
                    'address' : "0x1C",
                    'type'    : "Status" },
                   {'name'    : "LastTxOpCode",
                    'address' : "0x1C",
                    'size'    : "4",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "RxOpCodeStatus",
                    'address' : "0x1D",
                    'type'    : "Status" },
                   {'name'    : "LastRxOpCode",
                    'address' : "0x1D",
                    'size'    : "4",
                    'lsBit'   : "0",
                    'type'    : "Status" },
                   {'name'    : "TxOpCodeCount",
                    'address' : "0x1E",
                    'type'    : "Status" },
                   {'name'    : "RxOpCodeCount",
                    'address' : "0x1F",
                    'type'    : "Status" },
 ],
     'commands': [{ 'name'   : "ResetRx"},
                  { 'name'   : "Flush"} ]
   }

File: cpswAsynPortDriverApp/deviceLibrarySrc/Pgp2bAxi.cc
#include "Pgp2bAxi.h"

#define addrSize 4

Pgp2bAxi IPgp2bAxi::create(const char *name)
{
Pgp2bAxiImpl v = CShObj::create<Pgp2bAxiImpl>(name);
Field f;
        f = IIntField::create("CountReset", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x00*addrSize);
        f = IIntField::create("ResetRx", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x01*addrSize);
        f = IIntField::create("Flush", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x02*addrSize);
        f = IIntField::create("Loopback", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x03*addrSize);
        f = IIntField::create("LocData", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x04*addrSize);
        f = IIntField::create("AutoStatus", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x05*addrSize);
        f = IIntField::create("Status", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxPhyReady", 1, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("TxPhyReady", 1, false, 1);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxLocalLinkReady", 2, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxRemLinkReady", 3, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("TxLinkReady", 4, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxLinkPolarity", 2, false, 8);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxRemPause", 4, false, 12);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("TxLocPause", 4, false, 16);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxRemOverflow", 4, false, 20);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("TxLocOverflow", 4, false, 24);
        v->CMMIODevImpl::addAtAddress( f , 0x08*addrSize);
        f = IIntField::create("RxRemLinkData", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x09*addrSize);
        f = IIntField::create("RxCellErrorCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x0A*addrSize);
        f = IIntField::create("RxLinkDownCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x0B*addrSize);
        f = IIntField::create("RxLinkErrorCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x0C*addrSize);
        f = IIntField::create("RxRemOverflow0Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x0D*addrSize);
        f = IIntField::create("RxRemOverflow1Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x0E*addrSize);
        f = IIntField::create("RxRemOverflow2Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x0F*addrSize);
        f = IIntField::create("RxRemOverflow3Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x10*addrSize);
        f = IIntField::create("RxFrameErrorCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x11*addrSize);
        f = IIntField::create("RxFrameCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x12*addrSize);
        f = IIntField::create("TxLocOverflow0Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x13*addrSize);
        f = IIntField::create("TxLocOverflow1Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x14*addrSize);
        f = IIntField::create("TxLocOverflow2Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x15*addrSize);
        f = IIntField::create("TxLocOverflow3Count", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x16*addrSize);
        f = IIntField::create("TxFrameErrorCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x17*addrSize);
        f = IIntField::create("TxFrameCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x18*addrSize);
        f = IIntField::create("RxClkFreq", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x19*addrSize);
        f = IIntField::create("TxClkFreq", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1A*addrSize);
        f = IIntField::create("TxOpCodeStatus", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1C*addrSize);
        f = IIntField::create("LastTxOpCode", 4, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1C*addrSize);
        f = IIntField::create("RxOpCodeStatus", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1D*addrSize);
        f = IIntField::create("LastRxOpCode", 4, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1D*addrSize);
        f = IIntField::create("TxOpCodeCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1E*addrSize);
        f = IIntField::create("RxOpCodeCount", 32, false, 0);
        v->CMMIODevImpl::addAtAddress( f , 0x1F*addrSize);
        return v;
}

CPgp2bAxiImpl::CPgp2bAxiImpl(Key &key, const char *name) : CMMIODevImpl(key, name, 0x00010000, LE)
{
}

